src/pudl/helpers.py

```diff
@@ -339,6 +339,229 @@ def is_doi(doi):
+            E.g. if a monthly reported dataframe is being merged onto a daily reported
+            dataframe, then the merge would be performed on ``["year", "month"]``.
+        merge_type: How the dataframes should be merged.
+            Values are ["left", "right", "outer", "inner", "cross"].
+            See :func:`pandas.DataFrame.merge`.
+        report_at_start: Whether the data in the dataframe whose report date is not being
+            kept in the merged output (in most cases the less frequently reported dataframe)
+            is reported at the start or end of the time period e.g. January 1st
+            for annual data.
+        kwargs : Additional arguments to pass to :func:`pandas.DataFrame.merge`.
+
+
+    Returns:
+        Merged contents of left and right input dataframes.
+
+    Raises:
+        ValueError: if ``left_date_col`` or ``right_date_col`` columns are missing from their
+            respective input dataframes.
+        ValueError: if any of the labels referenced in ``shared_merge_cols`` are missing from either
+            the left or right dataframes.
+
+    """
+
+    def separate_date_cols(df, date_col_name):
+        df[date_col_name] = pd.to_datetime(df[date_col_name])
+        df.loc[:, "year"] = df[date_col_name].dt.year
+        df.loc[:, "quarter"] = df[date_col_name].dt.quarter
+        df.loc[:, "month"] = df[date_col_name].dt.month
+        df.loc[:, "day"] = df[date_col_name].dt.day
+        df = df.drop(date_col_name, axis=1)
+        return df
+
+    right = right.copy()
+    left = left.copy()
+    right = separate_date_cols(right, right_date_col)
+    left = separate_date_cols(left, left_date_col)
+    merge_cols = temporal_merge_cols + shared_merge_cols
+    out = left.merge(right, on=merge_cols, how=merge_type, **kwargs)
+
+    # reconstruct the report_date column and do some column cleanup
+    # Note: not sure if this works for "cross" merge
+    if "suffixes" in kwargs:
+        suffixes = kwargs["suffixes"]
+    else:
+        suffixes = ["_x", "_y"]
+    cols_to_drop = ["year", "month", "day"]
+    if "quarter" in out:
+        cols_to_drop.append("quarter")
+    else:
+        cols_to_drop += ["quarter" + suffixes[0], "quarter" + suffixes[1]]
+    for col in ["year", "month", "day"]:
+        if col not in out:
+            left_right_cols = [col + suffixes[0], col + suffixes[1]]
+            if report_at_start:
+                out.loc[:, col] = out[left_right_cols].max(axis=1)
+            else:
+                out.loc[:, col] = out[left_right_cols].min(axis=1)
+            cols_to_drop += left_right_cols
+    out.insert(
+        loc=0, column=new_date_col, value=pd.to_datetime(out[["year", "month", "day"]])
+    )
+    out = out.drop(cols_to_drop, axis=1)
+
+    return out
+
+
+def expand_timeseries(
+    df: pd.DataFrame,
+    date_col: str = "report_date",
+    start: str = None,
+    end: str = None,
+    freq="MS",
```

Comment:
> Should `freq` be restricted to a particular set of literals? Can we get that set from `timeseries.offset_aliases`?

Since you can combine frequencies and still have a valid string ("5day5min") it's hard to restrict `freq` to a set of literals. I instead try/except with the pandas function `to_offset(freq)` which will generate an exception if `freq` is not valid.

> It seems like this function implicitly assumes that all dates will be "start" format -- beginning of the month/year etc. Might want to state that explicitly. E.g. if you only do `ffill` to fill in the data values it'll work as expected one `MS` data, but it'll do nothing with month-end data (which would require `bfill`)

I believe it still works with "end" format dates in certain cases. Take the following input for example with `frequency = "M"` and a date range from `2019-01-01` to `2019-05-01`: 
```
plant_id    report_date    data
1           2019-01-31       1
1           2019-04-30       2 
```
This would `ffill` to 
```
plant_id    report_date    data
1           2019-01-31       1
1           2019-02-28       1
1           2019-03-31       1
1           2019-04-30       2 
1           2019-05-31       2 
```
and would `bfill` to
```
plant_id    report_date    data
1           2019-01-31       1
1           2019-02-28       2
1           2019-03-31       2
1           2019-04-30       2 
1           2019-05-31       Nan 
```
It depends on the use case but I think `ffill` is more intuitive to me.

There's also the case where you'd want to use `bfill`, but I think this is handled already in `date_merge`. For example, if you have annual data reported on December, 31st being merged with monthly data. Specifying `report_at_start = False` in `date_merge` would result in the newly merged on annual records being backfilled for every month in that year. If some months were not observed in the monthly dataset, then you could then call `expand_timeseries` to fill in these missing months.

Reply:
> Also maybe want to alert the user to the need to explicitly indicate the end date at the end of the year if they're working with annual data... like you want `start="2019-01-01"` and `end="2020-12-31"` -- the default args would get the start date right, but only take you up to January in the last year. Right?
> 
In this case since the data is annual wouldn't you only want data for `2019-01-01` and `2020-01-01`, not for `2021-01-01`? And if the data was monthly instead of annual you'd get data through December, 2020. 

> And some warning against using this function for irregularly reported data. It can turn annual data into corresponding homogeneous monthly data, but if you apply it to something like the `fuel_receipts_costs_eia923` table you'll get meaningless output.

Yeah, irregularly reported data is an issue unless you had a very granular frequency, like daily or something. But this seems like more of user error. For example, if you had data reported on the following dates:
```
2019-01-06
2019-02-03
2019-02-06
```
then it seems like you wouldn't want to actually give it a frequency of `M` or `MS`, since you probably want it at a daily resolution? If your dates were at a regular interval but on the 6th day of the month or something, you can still pass a frequency string corresponding to monthly starting on the 6th day.